kong/tools/stream_api.lua

```diff
@@ -0,0 +1,100 @@
+
+require "lua_pack"
+
+
+local kong       = kong
+local st_pack    = string.pack      -- luacheck: ignore string
+local st_unpack  = string.unpack    -- luacheck: ignore string
+local st_format  = string.format
+
+local PREFIX = ngx.config.prefix()
+
+local stream_api = {}
+
+local _handlers  = {}
+
+
+function stream_api.load_handlers()
+  local utils = require "kong.tools.utils"
+
+  for plugin_name in pairs(kong.configuration.loaded_plugins) do
+    local loaded, custom_endpoints = utils.load_module_if_exists("kong.plugins." .. plugin_name .. ".api")
+    if loaded and custom_endpoints._stream then
+      kong.log.debug("Register stream api for plugin: ", plugin_name)
+      _handlers[plugin_name] = custom_endpoints._stream
+      custom_endpoints._stream = nil
+    end
+  end
+end
+
+function stream_api.request(key, data, socket_path)
+  if type(key) ~= "string" or type(data) ~= "string" then
+    kong.log.error("Key and data must be strings")
+    return
+  end
+  if #data > 8000 then
+    kong.log.error("too much data")
```

Comment:
I still feel that we should provide a stronger guarantee in regard of the maximum length supported by throwing a Lua error, otherwise people may be tempting to send large amount of data through this channel which is not designed for it.

Just logging the error would be much less useful for the end user.

Reply:
done